### PR TITLE
Unify algorithm settings in our test suite

### PR DESCRIPTION
**Pull request by hanlint**
Several parts of our test suite need to test the algorithm within a trainer context:
* `test_algorithm_resumption`
* `test_trainer`: `TestTrainerAssets`
* individual algorithm tests

Currently, each one defines a large test combination of models, datasets, algorithms, and then filters/xfails, and each one takes a different approach. Algorithm developers would then need to modify these individually. This inflexiblity also slows the test suite as ReNet50 is used for `blurpool` because its required for `stochastic_depth`, when a more minimal model is appropriate.

This PR introduces a central place for storing canonical "setting" (dataset, model, algorithm, algorithm kwargs) for each algorithm. This should be as minimal as possible to activate the algorithms.

Draft mode, open for comments.

Open question: Should this be in code, or do we want to involve `hparams` and use the `yamls`? 

TODO:
- [x] adapt test_trainer to use settings
- [x] adapt test_algo_resumption to use settings

### CODE REVIEW

**Review by ravi-mosaicml — COMMENTED** [review]
I like the idea of putting the algorithm settings in one place.

However, I'm not a huge fan of removing all the YAHP from the test cases -- I would rather we use the AlgorithmHpraram / TrainerHparam classes. My main motiviation for using the YAHP classes is it allows for pyright to catch any typeerrors / name errors in tests (before actually running them), and otherwise we will slowly lose test coverage for the hparam classes.

**Comment by hanlint** [comment]
I try to make sure that unit tests only test one component, and be as independant as possible, so mixing hparams testing and algorithm resumption testing seems anathema to that? The motivation here being that the tests are more readable, easier to maintain, and also easier to debug (e.g. an hparams failure shows up in an hparams test, and not elsewhere).

Perhaps we need separate hparams test that create the object using both hparams and `__init__`, and then test that the created objects are equivalent?

File: tests/algorithms/test_algorithm_resumption.py [hunk 1]
```diff
@@ -101,157 +26,81 @@
...
-    if algo_name in skiplist[model_name].keys():
-        pytest.skip(skiplist[model_name][algo_name])
-
-    max_epochs = "5ep"
-    subset_num_batches = 5
-
-    trainer_hparams = TrainerHparams(algorithms=[],
-                                     seed=seed,
-                                     optimizer=AdamHparams(),
-                                     schedulers=[ExponentialSchedulerHparams(gamma=0.9)],
```
Comment (growlix) [inline tests/algorithms/test_algorithm_resumption.py:142]: If you're using the default scheduler hparams, just make sure that gamma is >> 0.1. If gamma is sufficiently small, the learning rate can decay so quickly that the model stops being updated, causing the checkpointing tests to pass (possibly) erroneously.

**Comment by growlix** [comment]
> tests are more readable, easier to maintain, and also easier to debug

Certainly seems that way to me.

> Perhaps we need separate hparams test that create the object using both hparams and __init__, and then test that the created objects are equivalent?

This seems like the right call.

**Review by growlix — COMMENTED** [review]
Review presented as individual comments above

File: tests/algorithms/test_algorithm_resumption.py [hunk 2]
```diff
@@ -101,157 +26,81 @@
...
+        'train_dataloader': DataLoader(dataset=setting['dataset'], batch_size=4),
+        'max_duration': '5ep',
+        'loggers': [],
+        'seed': seed,
+        'device': device,
+        'deterministic_mode': True,
+        'save_filename': save_filename,
+        'save_folder': folder1,
+        'save_interval': save_interval,
+        'train_subset_num_batches': 5,
```
Comment (ravi-mosaicml) [inline tests/algorithms/test_algorithm_resumption.py:69]: Can this be shorter? Say 2?

File: tests/algorithms/test_algorithm_resumption.py [hunk 3]
```diff
@@ -101,157 +26,81 @@
...
+
+def _assert_checkpoints_equal(file1, file2):
+    checkpoint1 = torch.load(file1)
+    checkpoint2 = torch.load(file2)
+
+    # compare rng
+    deep_compare(checkpoint1['rng'], checkpoint2['rng'])
+
+    # compare state
+    _check_dict_recursively(checkpoint1['state'], checkpoint2['state'])
```
Comment (ravi-mosaicml) [inline tests/algorithms/test_algorithm_resumption.py:106]: `tests.common` includes a `assert_state_equivalent`.

(This function should probably be split up into `assert_state_equivalent` and `assert_state_dict_equivalent`)



```suggestion
    assert_state_dict_equivalent(checkpoint1['state'], checkpoint2['state'])
```

**Review by ravi-mosaicml — COMMENTED** [review]
I'm sold on removing hparams from tests, and testing the hparam classes separately.

Overall logic seems fine, though main comments would be to please try to preserve typing in tests (and not us kwargs unless really necessary). Static analysis tools, such as VSCode's refactoring and pyright, cannot automatically read through kwargs. While I realize that explicit naming would result in more test code, I would prefer it, as I think it helps with making refactoring less painful and tests more readable.

**Review by ravi-mosaicml — APPROVED** [review]

File: tests/algorithms/algorithm_settings.py [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+    'model': (ComposerResNet, {
+        'model_name': 'resnet18',
+        'num_classes': 2
+    }),
+    'dataset': (common.RandomImageDataset, {
+        'shape': (3, 224, 224),
+    }),
+}
+
+_settings = {
```
[the remark this answers is not in the document]
Comment (hanlint) [reply tests/algorithms/algorithm_settings.py:35]: I would prefer the key be the same as what is registered in the `algorithms_registry`, so that the retrieval for both is the same.

File: tests/algorithms/algorithm_settings.py [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+            (obj, kwargs) = setting[key]
+        else:
+            (obj, kwargs) = (setting[key], {})
+
+        # create the object
+        result[key] = obj(**kwargs)
+
+    # create algorithm
+    kwargs = setting.get('kwargs', {})
+    hparams = algorithm_registry.get_algorithm_registry()[name]
```
[the remark this answers is not in the document]
Comment (hanlint) [reply tests/algorithms/algorithm_settings.py:131]: Hmm I would much prefer to construct the algorithm directly as you suggest, but not sure where to get a list of all the class objects. Short of maintaining a separate list from the `algorithms_registry`, which would be duplicative, not sure of an approach here, but I'm open to ideas for a follow up PR.

File: tests/algorithms/test_algorithm_registry.py [hunk 6]
```diff
@@ -11,6 +12,8 @@
                                  SelectiveBackpropHparams, SeqLengthWarmupHparams, SqueezeExciteHparams,
                                  StochasticDepthHparams, SWAHparams, algorithm_registry)
 from composer.core.algorithm import Algorithm
+from composer.models.base import ComposerModel
+from tests.algorithms.algorithm_settings import get_settings
 
 default_required_fields = {
```
[the remark this answers is not in the document]
Comment (hanlint) [reply tests/algorithms/test_algorithm_registry.py:18]: good catch, simplified significantly!

File: tests/algorithms/test_algorithm_resumption.py [hunk 7]
```diff
@@ -101,157 +26,81 @@
...
+        pytest.xfail('Known issues')
+
+    setting = get_settings(algorithm)
+    if setting is None:
+        pytest.skip('No setting provided in algorithm_settings.')
+
+    folder1 = os.path.join(tmpdir, 'folder1')
+    folder2 = os.path.join(tmpdir, 'folder2')
+
+    config = {
```
[the remark this answers is not in the document]
Comment (hanlint) [reply tests/algorithms/test_algorithm_resumption.py:57]: For unit tests, I think readability (e.g. understanding what has changed between trainer1 and trainer2), trumps any pyright.

File: tests/algorithms/test_algorithm_resumption.py [hunk 8]
```diff
@@ -101,157 +26,81 @@
...
+    if algorithm in ('cutmix, mixup, label_smoothing'):
+        # see: https://github.com/mosaicml/composer/issues/362
+        pytest.importorskip("torch", minversion="1.10", reason="Pytorch 1.10 required.")
+
+    if algorithm in ('layer_freezing', 'swa', 'stochastic_depth'):
+        pytest.xfail('Known issues')
+
+    setting = get_settings(algorithm)
+    if setting is None:
+        pytest.skip('No setting provided in algorithm_settings.')
```
[the remark this answers is not in the document]
Comment (hanlint) [reply tests/algorithms/test_algorithm_resumption.py:52]: done!

File: tests/algorithms/test_algorithm_resumption.py (hunk 3 above)
Comment (hanlint) [reply tests/algorithms/test_algorithm_resumption.py]: I switched this to use `deep_compare`, which seems to pass the tests with `atol=0, rtol=0` on GPU. This is stricter than the `assert_state_dict_equivalent`, which has a pretty worrisome tolerances:

```
    atol = 0.1 if is_gpu else 0.0
    rtol = 0.1 if is_gpu else 0.0
```
FYI, we've seen very reproducible ResNet50 runs, where the same seed yields essentially the same accuracy over 90 epochs of training, so we should definitely trace where this is coming from.

File: tests/algorithms/test_algorithm_resumption.py (hunk 1 above)
Comment (hanlint) [reply tests/algorithms/test_algorithm_resumption.py:142]: good suggestion! I added a control test to make sure that the model weights are indeed being updated.
